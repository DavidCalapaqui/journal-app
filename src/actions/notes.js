import { db } from "../firebase/firebase-config";
import { collection, addDoc, updateDoc, doc, deleteDoc } from "firebase/firestore";
import {types} from '../types/types';

import { loadNotes } from "../helpers/loadNotes";
import Swal from "sweetalert2";
import { fileUpload } from "../helpers/fileUpload";



export const startNewNote = () =>{
    
    
    return async (dispatch, getState) =>{
        const {uid} = getState().auth;
        
        const newNote = {
            title: '',
            body: '',
            date: new Date().getTime()
        }

        try {
            const doc = await addDoc(collection(db, `${uid}/journal/notes`), newNote);
            //console.log('DOC REF: ', doc);

            dispatch(activeNote(doc.id, newNote));
            dispatch(addNewNote(doc.id, newNote));
        } catch (error){
            console.error('ERROR AL INTENTAR GRABAR UNA COLECCION: ',error);
        } 
       
    }
}

export const startLoadingNotes = (uid) =>{
    return async(dispatch)=>{
        const notes = await loadNotes(uid);
        dispatch(setNotes(notes));
    }
}

export const activeNote = (id, note) =>({
    type: types.notesActive,
    payload:{
        id, 
        ...note
    }
})

export const addNewNote = (id, note)=>({
    type: types.notesAddNew,
    payload:{
        id,
        ...note
    }
})

export const setNotes = (notes) =>({
    type: types.notesLoad,
    payload: notes
})


export const startSaveNote = (note) =>{
    return async (dispatch, getState)=>{
        const {uid} = getState().auth;

        if(!note.url){
            delete note.url;
        }

        const noteToFirestore = {...note};
        delete noteToFirestore.id;
        const noteRef = doc(db, `${uid}/journal/notes/${note.id}`);
        try {
             await updateDoc(noteRef, noteToFirestore);
        } catch (error) {
            console.error('ERROR AL ACTUALIZAR NOTA: ',error);
        }

        dispatch(refreshNote(note.id, note));
        Swal.fire('Saved', note.title, 'success')
    }
}

export const refreshNote = (id, note) =>({
    type: types.notesUpdated,
    payload:{
        id, 
        note:{
            id,
            ...note
        }
    }
})


export const startUploading = (file) =>{

    return async (dispatch,getState ) =>{
        const {active:activeNote} = getState().notes;
        Swal.fire({
            title: 'Uploading',
            text: 'Please wait...',
            allowOutsideClick: false,
            onBeforeOpen: ()=>{ 
                Swal.showLoading();
            }
        });
        const fileUrl = await fileUpload(file);
        activeNote.url = fileUrl;
        dispatch(startSaveNote(activeNote))
        Swal.close();
    }
}


export const startDeleting = (id) =>{
    return async (dispatch, getState)=>{
        const uid = getState().auth.uid;

        const noteRef = doc(db, `${uid}/journal/notes/${id}`);

        try {
            await deleteDoc(noteRef);
            dispatch(deleteNote(id));
            Swal.fire('Note removed', 'The note has been removed on database', 'success')

        } catch (error) {
            Swal.fire('Error', error.message, 'error')
        }

    }
}

export const deleteNote = (id) =>({
    type: types.notesDelete,
    payload: id
})

//limpiar las notas del usuario en sesión cuando cierre sesión
export const notesLogout =()=>({
    
    type: types.notesLogoutCleaning
})