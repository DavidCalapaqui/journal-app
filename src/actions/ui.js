import { types } from "../types/types";

//ACCIONES PARA ERRORES EN CAMPOS
export const setError = (err) =>({
    type: types.uiSetError,
    payload: err
})

export const removeError = () =>({
    type: types.uiRemoveError,
})

//ACCIONES PARA CARGANDO
export const startLoading = () =>({
    type: types.uiStartLoading,
})

export const finishLoading = () =>({
    type: types.uiFinishLoading,
})