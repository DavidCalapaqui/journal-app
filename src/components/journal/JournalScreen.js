import React from 'react'
import { useSelector } from 'react-redux'
import { NoteScreen } from '../notes/NoteScreen'
import { NothingSelected } from './NothingSelected'
import { Sidebar } from './Sidebar'

//import { NothingSelected } from './NothingSelected'
export const JournalScreen = () => {
    

    //jalo el active del state.notes
    const {active} = useSelector(state => state.notes);

    
    //si tengo una nota seleccionada la muestro si no la pantalla de sleccione una
    return (
        <div className='journal__main-content animate__animated animate__fadeIn animate__faster'>
            <Sidebar />
        
            <main>
                {
                    (active)
                    ? <NoteScreen />
                    : <NothingSelected />
                }
                             
            </main>
        </div>
    )
}
