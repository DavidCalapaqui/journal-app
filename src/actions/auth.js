
import {getAuth, signInWithPopup, createUserWithEmailAndPassword, updateProfile, signInWithEmailAndPassword} from 'firebase/auth';
import { googleAuthProvider } from '../firebase/firebase-config';
import Swal from 'sweetalert2';

import { types } from "../types/types";
import { finishLoading, startLoading } from './ui';
import { notesLogout } from './notes';

//accion
export const startLoginEmailPassword = (email, password) =>{
    //retorna un callback 
    return (dispatch)=>{
        //dispatch startLogin
        dispatch(startLoading());

        const auth = getAuth();
        return signInWithEmailAndPassword(auth, email, password)
        .then(async ({user})=>{
            dispatch(login(user.uid, user.displayName));
            
            dispatch(finishLoading());
            //dispatch stopLoggin
            //console.log('EMAIL & PASS - USER SIGIN: \n',user.displayName, user.uid);
        }).catch(e=>{
            Swal.fire('Error', e.message, 'error' );
            }
        );
    }

}

export const startRegisterWithEmailPasswordName = (email, password, name) =>{
    return(dispatch)=>{
        const auth = getAuth();
        createUserWithEmailAndPassword(auth, email, password)
        .then(async({user})=>{ 
            //dispatch(login(user.uid, user.displayName));
            await updateProfile(user, {displayName: name});
            dispatch(user.uid, user.displayName);
            console.log(user);
        })
        .catch(e=>{
            Swal.fire('Error', e.message, 'error' );
        })
    }

}

//accion
export const startGoogleLogin = () =>{
    return (dispatch)=>{
       const auth = getAuth();
       signInWithPopup(auth, googleAuthProvider)
       .then(({user})=>{
            console.log('GOOGLE - USER SIGIN: \n', user.displayName, user.uid);
           dispatch(login(user.uid, user.displayName));
       })
    }
}
//accion
export const login = (uid, displayName) =>{
    return {
        type: types.login,
        payload: {
            uid,
            displayName
        }
    }
}

//acción asíncrona a firebase
export const startLogout = () =>{
    const auth = getAuth();

    return async (dispatch)=>{
        await auth.signOut();
        dispatch(logout()); 
        //disparar accion para limpiar notas
        dispatch(notesLogout());
    }
}

export const logout = () =>({
    type: types.logout
});