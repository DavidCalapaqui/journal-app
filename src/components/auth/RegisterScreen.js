import React from 'react'
import { useDispatch, useSelector } from 'react-redux';
import { Link } from 'react-router-dom'
import validator from 'validator';
import { startRegisterWithEmailPasswordName } from '../../actions/auth';
import { setError, removeError } from '../../actions/ui';


import { useForm } from '../../hooks/useForm'


export const RegisterScreen = () => {

    const dispatch = useDispatch();
    
    const {msgError} = useSelector(state => state.ui);
    
    
    //obtener info de los campos del formulario
    const [formValues, handleInputChange] = useForm({
        name:"David", 
        email:'david.calapaqui98@gmail.com', 
        password:'123456', 
        password2:'123456'

    });

    const {name, email, password, password2} = formValues;

    const handleRegister = (e)=>{
        e.preventDefault();

        if(isFormValid()){
            dispatch(startRegisterWithEmailPasswordName(email, password, name));
        }
        
    }

    const isFormValid = () =>{
       if(name.trim().length ===0){
           dispatch(setError('Name is required'));
           return false;
       }else if(!validator.isEmail(email)){
           dispatch(setError('Invalid email'));   
           return false;
       }else if(password !== password2 || password.length <5){
         dispatch(setError('Password should be at least 6 characters and match each other'));  
         return false
       }
       
       dispatch(removeError())
       return true;
    }
    
    return (
        <>
        <h3 className='auth__title'>Register</h3>

        <form onSubmit={handleRegister}
            className="animate__animated animate__fadeIn animate__faster"
        > 
            
            {    
                msgError &&
                (<div className='auth__alert-error'>
                    {msgError}
                </div>
                ) 
            }
            
            <input
                type="text"
                placeholder='Name'
                name="name"
                className='auth__input'
                autoComplete='off'
                value={name}
                onChange={handleInputChange}
            />
            <input
                type="text"
                placeholder='Email'
                name="email"
                className='auth__input'
                autoComplete='off'
                value={email}
                onChange={handleInputChange}
            />
           <br/>
            <input
                type="password"
                placeholder="Password"
                name="password"
                className='auth__input'
                value={password}
                onChange={handleInputChange}
            />
            <input
                type="password"
                placeholder="Confirm password"
                name="password2"
                className='auth__input'
                value={password2}
                onChange={handleInputChange}
            />
            
            <button
                className='btn btn-primary btn-block mb-5' 
                type="submit"
                
                
            >Register</button>

            <hr />

                   

            <Link 
            className='link'
            to="/auth/login">

               Already registered?
            </Link>

        </form>

    </>
    )
}
