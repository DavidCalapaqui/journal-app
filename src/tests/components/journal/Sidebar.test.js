import React from 'react';
import {mount} from 'enzyme';
import {Provider} from 'react-redux';
import configureStore from 'redux-mock-store'; //ES6 modules
import thunk from 'redux-thunk';
//import {MemoryRouter} from 'react-router-dom'; //finjir rutas

import { Sidebar } from '../../../components/journal/Sidebar';
import { startLogout } from '../../../actions/auth';
import { startNewNote } from '../../../actions/notes';

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

jest.mock('../../../actions/auth', ()=>({
    startLogout: jest.fn(),
}))
jest.mock('../../../actions/notes', ()=>({
    startNewNote: jest.fn(),
}))


const initState = {
    auth: {
        uid: 'GcuFhm6rFWeVQm9bfJpFQoqiQ9C2',
        name: 'David'
    },
    ui: {
        loading: false,
        msgError: null
    },
    notes:{
        notes: [
            {
                id    :"ENqiu5vFclsHuA5geru3",
                title :"Note title!!!",
                date  :1644273441664,
                body  :"This is anther note!!!",
                url   :"https://res.cloudinary.com/api-node-cloud/image/upload/v1644363562/journal-photos/r4rn9t0sczxtzhrbewfl.png"
            }
        ]
    }
}

//simulacion del store
let store = mockStore(initState);
//reemplazar la funcionalidad del store por una funcion de jest para poder evaluarla
store.dispatch = jest.fn();

const wrapper = mount(
    
    <Provider store={store}>
            <Sidebar/>
    </Provider>
        
);




describe('Pruebas en el Sidebar', ()=>{

    test('Debe de mostrarse correctamente', () => { 
        expect(wrapper).toMatchSnapshot();    
    })

    test('Debe de llamar la acccion startLogout', () => { 
        wrapper.find('.btn').prop('onClick')();
        expect(startLogout).toHaveBeenCalled();

    })

    test('Debe de llamar el startNewNote', () => { 
        wrapper.find('.journal__new-entry').prop('onClick')();
        expect(startNewNote).toHaveBeenCalled();   

    })
})