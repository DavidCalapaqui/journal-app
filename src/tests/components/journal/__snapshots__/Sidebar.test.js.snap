// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el Sidebar Debe de mostrarse correctamente 1`] = `
<aside
  className="journal__sidebar"
>
  <div
    className="journal__sidebar-navbar"
  >
    <h3
      className="mt-5"
    >
      <i
        className="far fa-moon"
      />
      <span>
         
        David
      </span>
    </h3>
    <button
      className="btn"
      onClick={[Function]}
    >
      Logout
    </button>
  </div>
  <div
    className="journal__new-entry"
    onClick={[Function]}
  >
    <i
      className="far fa-calendar-plus fa-5x"
    />
    <p
      className="mt-5"
    >
      New Entry
    </p>
  </div>
  <div
    className="journal__entries animate__animated animate__shakeY"
  >
    <div
      className="journal__entry pointer"
      onClick={[Function]}
    >
      <div
        className="journal__entry-picture"
        style={
          Object {
            "backgroundImage": "url(https://res.cloudinary.com/api-node-cloud/image/upload/v1644363562/journal-photos/r4rn9t0sczxtzhrbewfl.png)",
            "backgroundSize": "cover",
          }
        }
      />
      <div
        className="journal__entry-body"
      >
        <p
          className="journal__entry-title"
        >
          Note title!!!
        </p>
        <p
          className="journal__entry-content"
        >
          This is anther note!!!
        </p>
      </div>
      <div
        className="journal__entry-date-box"
      >
        <span>
          Monday
        </span>
        <h4>
          7th
        </h4>
      </div>
    </div>
  </div>
</aside>
`;
